### DISCUSSION TITLE
Class-based Components

### DISCUSSION THREAD

**Discussion by loreanvictor** [loreanvictor/callbag-jsx#14] [Q&A]
> Are there any plans to support classes as components when rendering the JSX? All examples seem to use a function that gets the props and renderer and returns what should be rendered. Have you considered supporting classes as well? For example, if the class has a method "render" it could be called with the props and renderer arguments, maybe. Some people might prefer classes over function for organizing their sources...
>
> This is not a big deal, since it should be easy to create a wrapper for such classes (so that we could do something like this after defining the class `const ComponentRenderer = wrapper(ComponentClass);`, but it might be interesting to support it out of the box...

_Originally posted by @rosenfeld in https://github.com/loreanvictor/callbag-jsx/issues/9#issuecomment-812222252_

**Comment by rosenfeld** [comment]
I just found this: https://loreanvictor.github.io/callbag-jsx/in-depth/renderer

There's a section about a plugin to allow for a component-based approach.

**Comment by loreanvictor** [comment]
> I just found this: https://loreanvictor.github.io/callbag-jsx/in-depth/renderer
> 
> There's a section about a plugin to allow for a component-based approach.

I was writing about the same thing :)) Yes, you can extend behavior of `renderer`s via plugins, for [example](https://loreanvictor.github.io/callbag-jsx/in-depth/renderer#example-class-based-components) a plugin can provide support for class based components.

I think it would also be a good idea to publish a package for such a plugin, but for now I don't see any benefits in providing the feature out of the box. I generally believe a bunch of integrating but independently maintained libraries will remain more agile and will be managed more efficiently (for example its easier to make changes to small libraries, its easier for new contributors to start contributing if the library is small, etc).

**Comment by rosenfeld** [comment]
Yes, but for tutorials/documentation purposes, I don't see component authors showing examples with class-based components unless they are supported out-of-the-box, for example...

**Comment by loreanvictor** [comment]
I think component authors would need to specifically provide each type of component, so I suspect IF a component author also provides a class-based version of a component, they would document it as well. It is notable that this would of course require a _standard_ class-based plugin that any component author could rely on.

Edit: actually I don't think there would be any particular issue for documentation purposes, since class-based components would be mostly consumed exactly the same way as function-based components, which means there wouldn't be any need for specific class-based documentation / tutorials on part of component authors.

**Reply by rosenfeld** [reply]
I meant that authors won't show examples using classes, I wasn't referring to the component usage itself, but on full app examples demonstrating the components. This isn't specific to component authors but also applies to generic articles / documentation... The lack of a standard to support classes in practice discourage its usage.

**Reply by loreanvictor** [reply]
Ok I better understand what you mean. I agree, if _class-based components_ do provide an actual benefit (beyond a stylistic preference) for particular situations, I think it would be a good idea to further encourage using the pattern.

However, if it is only a stylistic change, I guess it would only lead to further confusion and unnecessary divergence in code-style. Atm, I can _vaguely_ imagine the following two scenarios where the pattern would be more beneficial:

- _The component has lots of callbacks, which are also complex_
  On this front, I am torn. In few on-production projects that I have done with callbag-jsx, the best solution to these situations seems to be further breaking the components down to smaller ones, as large classes are still not that much cleaner than large functions with lots of callbacks defined inside them.
  
- _The component would want to provide control methods to outer components_
  For example, imagine a dialog component that would want to provide a `.close()` method to outer-component (who would also be specifying the child nodes of this component). This I believe would be an actual use-case for the pattern, since otherwise this would need to be done via solutions similar to React's [`useImperativeHandle()`](https://reactjs.org/docs/hooks-reference.html#useimperativehandle), which are not really idea.
  
  However, in practice these scenarios seem to be too rare for that to be a serious issue, as most of the time the best way to communicate with a component is via its props. This means it would also provide yet another method of communicating with a component (alongside yet another way of defining a component), which would basically be the same confusion / unnecessary divergence issue but twice as bad.
  
Note that this assessment is based on my current experience with the library on serious projects. I do still suspect that in some cases class-based components would be a better option. Thats why there is a class-based plugin example in the docs, to encourage some experimentation with the pattern and get more actual examples / data to make a decision on.

**Reply by rosenfeld** [reply]
I've created a React app once for which almost all components would inherent from a base custom component that provided many utilities out of the box. The application supported moving back and forward while updating the url so that the specific state could be shared with others. Without classes, those helpers would require way more verbosity.

**Reply by loreanvictor** [reply]
Hmmm, yes inheritance can also be a useful concept for components, though it can sometimes lead to confusion about how the layouts are intermixed (though I suspect this wouldn't be much of a problem in callbag-jsx as child components can simply override and re-use parent `.render()` however they see fit).

The particular use-case you mention, sounds to me though, that would be better handled by a specific rendering plugin providing utilities to all components (or a separate _routing_ system). Still, it would be interesting to see some such architectures implemented in callbag-jsx as it stands and further reflect on how they could be improved by class-based components.

**Comment by loreanvictor** [comment]
Ok I'll move this to discussions for now.
